icu4c/source/i18n/unitsdata.cpp

```diff
@@ -53,116 +54,69 @@ class ConversionRateDataSink : public ResourceSink {
+            ConversionRateInfo *cr = outVector.emplaceBack();
+            if (!cr) {
+                status = U_MEMORY_ALLOCATION_ERROR;
+                return;
+            } else {
+                cr->sourceUnit.append(srcUnit, status);
+                cr->baseUnit.appendInvariantChars(baseUnit, status);
+                cr->factor.appendInvariantChars(factor, status);
+                if (!offset.isBogus()) cr->offset.appendInvariantChars(offset, status);
+            }
         }
+        return;
     }
 
   private:
     MaybeStackVector<ConversionRateInfo> &outVector;
 };
 
-/**
- * Collects conversion information for a "SINGLE" unit (a unit whose complexity
- * is UMEASURE_UNIT_SINGLE). For a COMPOUND or SEQUENCE unit an error will
- * occur.
- *
- * @param unit The input unit. Its complexity must be UMEASURE_UNIT_SINGLE, but
- * it may have a dimensionality != 1.
- * @param converUnitsBundle A UResourceBundle instance for the convertUnits
- * resource.
- * @param convertSink The ConversionRateDataSink through which
- * ConversionRateInfo instances are to be collected.
- * @param status The standard ICU error code output parameter.
- */
-void processSingleUnit(const MeasureUnit &unit, const UResourceBundle *convertUnitsBundle,
-                       ConversionRateDataSink &convertSink, UErrorCode &status) {
-    if (U_FAILURE(status)) return;
-    int32_t dimensionality = unit.getDimensionality(status);
-
-    // Fetch the relevant entry in convertUnits.
-    MeasureUnit simple = unit;
-    if (dimensionality != 1 || simple.getSIPrefix(status) != UMEASURE_SI_PREFIX_ONE) {
-        simple = unit.withDimensionality(1, status).withSIPrefix(UMEASURE_SI_PREFIX_ONE, status);
-    }
-    ures_getAllItemsWithFallback(convertUnitsBundle, simple.getIdentifier(), convertSink, status);
-}
-
 } // namespace
 
-MaybeStackVector<ConversionRateInfo> U_I18N_API
-getConversionRatesInfo(const MaybeStackVector<MeasureUnit> &units, UErrorCode &status) {
+MaybeStackVector<ConversionRateInfo> U_I18N_API getAllConversionRates(UErrorCode &status) {
     MaybeStackVector<ConversionRateInfo> result;
-    if (U_FAILURE(status)) return result;
-
     LocalUResourceBundlePointer unitsBundle(ures_openDirect(NULL, "units", &status));
-    StackUResourceBundle convertUnitsBundle;
-    ures_getByKey(unitsBundle.getAlias(), "convertUnits", convertUnitsBundle.getAlias(), &status);
-    ConversionRateDataSink convertSink(result);
-
-    for (int i = 0; i < units.length(); i++) {
-        int32_t numSingleUnits;
-        LocalArray<MeasureUnit> singleUnits = units[i]->splitToSingleUnits(numSingleUnits, status);
-
-        for (int i = 0; i < numSingleUnits; i++) {
-            processSingleUnit(singleUnits[i], convertUnitsBundle.getAlias(), convertSink, status);
-        }
-    }
+    ConversionRateDataSink sink(result);
+    ures_getAllItemsWithFallback(unitsBundle.getAlias(), "convertUnits", sink, status);
     return result;
 }
 
+MaybeStackVector<ConversionRateInfo>
```

Comment:
Feedback from @sffc - the code we currently have (here in #31 and in #32 ) has a lot of "MaybeStackVector<ConversionRateInfo> &" being passed around. sffc suggests we make a first-class type called ConversionRateInfoList with the methods you need, that contains the MaybeStackVector inside.

To be reviewed together with #32's use-cases.

Reply:
My thoughts: I think returning a MaybeStackVector from getAllConversionRates might be reasonable enough. It could be packaged into a ConversionRateInfoList by other code? There is also a question of where to do the symbolic maths: but I think that might be best done when the converter is set up, if it's possible for the symbolic maths to cancel out? (Though if we don't care about maximum accuracy, we could simply deal with it in double precision maths.)

Reply:
The typical pattern is that you heap-allocate an object to store the results from a data loading operation, and use the sink to populate it.  I think you should do the same here.  I would rather err on the side of too many types than too few; I often add types where I didn't have types before, but very rarely the other direction.

For the time being, if it's easier, you can add a TODO in order to unblock #32.

Reply:
I first switched to passing in a MaybeStackVector as a pointer parameter, on the stylistic choice that it's clearer at the call-site that it's an output parameter.

However these (naive) grep's suggest references are more common in ICU:

    egrep '& result|&result' i18n/*.h
    egrep '\* result|\*result' i18n/*.h

so I've now changed it to references. PTAL?

Reply:
I would still rather see an extra type ConversionRateInfoList encapsulating the MaybeStackVector, but not strongly enough to block this PR.